Overhauls discord verification system

optimumtact opened the pull request:
This completely replaces the previous verification system, for one that
will interoperate with a discord redbot instance that uses the cogs
located at you github.com/optimumtact/orangescogs

This cuts out several steps in the system, but it also leaves alone the
existing notify system (which just uses a file list of discord ids) as a
record of who to notify

SQL changes required for the new database system

Version 5.10, 7 August 2020, by oranges

Changes how the discord verification process works.
Adds the discord_links table, and migrates discord id entries from
player table to the discord links table in a once off operation and then
removes the discord id on the player table

The user connects to any tg server, and uses the "Verify Discord
Account" verb, this generates a six word one time use token, with a 4
hour time validity period (defined as 4 hours from the timestamp value)
in the discord links table.

This one time token, and the ckey of the user are stored in
discord_links

At this point the entire DM side is done, this is all it does

optimumtact commented:
DM side is tested and working.

bot side has been ready to go for a while.
![](https://file.house/zlNE.png)

Jordie0608 commented on SQL/database_changelog.txt:
```diff
@@ -1,15 +1,42 @@
+	`ckey` VARCHAR(32) NOT NULL,
+	`discord_id` BIGINT(20) DEFAULT NULL,
+	`timestamp` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	`one_time_token` TEXT NOT NULL,
```
This was changed to `varchar(100)` in the schema.

Jordie0608 commented on SQL/tgstation_schema.sql:
```diff
@@ -574,6 +573,20 @@ END
+	`discord_id` BIGINT(20) DEFAULT NULL,
+	`timestamp` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	`one_time_token` VARCHAR(100) NOT NULL,
+  	`valid` BOOLEAN NOT NULL DEFAULT FALSE,
```
​

Jordie0608 commented on config/dbconfig.txt:
```diff
@@ -3,40 +3,40 @@
 
 ## Should SQL be enabled? Uncomment to enable
-#SQL_ENABLED
+SQL_ENABLED
```
You've accidentally committed this file.

Jordie0608 commented on code/controllers/subsystem/discord.dm:
```diff
@@ -81,70 +91,186 @@ SUBSYSTEM_DEF(discord)
+	var/timeboundsql = ""
+	if(timebound)
+		timeboundsql = "AND timestamp >= Now() - INTERVAL 4 HOUR"
+	var/query = "SELECT CAST(discord_id AS CHAR(25)), ckey, MAX(timestamp), one_time_token FROM [format_table_name("discord_links")] WHERE one_time_token = :one_time_token [timeboundsql] GROUP BY ckey, discord_id, one_time_token LIMIT 1"
```
What's this cast for?

optimumtact replied:
byond converts it to scientific notation otherwise

Jordie0608 commented on code/controllers/subsystem/discord.dm:
```diff
@@ -81,70 +91,186 @@ SUBSYSTEM_DEF(discord)
+		return
+	if(query_get_discord_link_record.NextRow())
+		var/result = query_get_discord_link_record.item
+		. = new /datum/discord_link_record(result[2], result[1], result[4], result[3])
```
Since this is only used in `generate_one_time_token()` to test the `while(not_unique)` loop, couldn't you just set `. = TRUE` and skip having to create a datum that isn't actually used for anything?

optimumtact replied:
it will be used in later code that lets people manage discord links in game.

Jordie0608 commented on code/controllers/subsystem/discord.dm:
```diff
@@ -81,70 +91,186 @@ SUBSYSTEM_DEF(discord)
+  * Remember, multiple entries can exist
+  *
+  * Arguments:
+  * * discordid The users discord id (string)
```
```suggestion
  * * discord_id The users discord id (string)
```

AffectedArc07 commented on config/dbconfig.txt:
```diff
@@ -3,40 +3,40 @@
 
 ## Password used to access the database.
-FEEDBACK_PASSWORD password
+FEEDBACK_PASSWORD tgstation
```
safe:tm:

optimumtact replied:
it's secure

optimumtact committed:
Overhauls discord verification system

This completely replaces the previous verification system, for one that
will interoperate with a discord redbot instance that uses the cogs
located at you github.com/optimumtact/orangescogs

This cuts out several steps in the system, but it also leaves alone the
existing notify system (which just uses a file list of discord ids) as a
record of who to notify

SQL changes required for the new database system

Version 5.10, 7 August 2020, by oranges

Changes how the discord verification process works.
Adds the discord_links table, and migrates discord id entries from
player table to the discord links table in a once off operation and then
removes the discord id on the player table

The user connects to any tg server, and uses the "Verify Discord
Account" verb, this generates a six word one time use token, with a 4
hour time validity period (defined as 4 hours from the timestamp value)
in the discord links table.

This one time token, and the ckey of the user are stored in
discord_links

At this point the entire DM side is done, this is all it does

LemonInTheDark commented:
Fuck it
